### PR TITLE
Server WMS temporal properties support

### PR DESCRIPTION
**Pull request by elpaso**
# Server WMS temporal properties

QGIS Server currently supports **Dimensions** ( `TIME`, `ELEVATION` etc.) for vector layers only and does not integrate with the (newer) temporal properties API of QGIS.

This PR exposes the temporal properties to the server: the server will expose the `TIME` dimension from the temporal properties to `GetCapabilities` and accept a `TIME` query string parameter to activate the time filtering.

## Accepted TIME syntax

The OAPIF implementation will be used to parse the time instants and ranges:

```
     * Interval syntax:
     *
     *   interval-closed     = date-time "/" date-time
     *   interval-open-start = [".."] "/" date-time
     *   interval-open-end   = date-time "/" [".."]
     *   interval            = interval-closed / interval-open-start / interval-open-end
     *   datetime            = date-time / interval

```

Note that a list of instants and ranges will not be supported at this time.

## Potential conflicts with defined TIME dimension

To solve the conflict, if a `TIME` dimension was explicitly defined for a vector layer, the temporal properties for that layer will not be activated and the explicitly set `TIME` dimension will be considered instead, this will solve compatibility issues with older projects.

## TODO

- [x] Tests
- [x] Decide how to handle open ending/beginning interval and how to expose them in the capabilities
- [x] Expose "Default" to capabilities

Funded by: Gis3W

### CODE REVIEW

File: src/core/vector/qgsvectorlayertemporalproperties.cpp [hunk 1]
```diff
@@ -442,11 +442,22 @@ QString dateTimeExpressionLiteral( const QDateTime &datetime )
          .arg( datetime.time().second() + datetime.time().msec() / 1000.0 );
 }
 
-QString QgsVectorLayerTemporalProperties::createFilterString( QgsVectorLayerTemporalContext, const QgsDateTimeRange &filterRange ) const
+QString QgsVectorLayerTemporalProperties::createFilterString( QgsVectorLayerTemporalContext context, const QgsDateTimeRange &filterRange ) const
 {
   if ( !isActive() )
     return QString();
 
+  auto dateTimefieldCast = [ &context ]( const QString & fieldName ) -> QString
```
Comment (elpaso) [inline src/core/vector/qgsvectorlayertemporalproperties.cpp:450]: CC @nyalldawson

Comment (nyalldawson) [reply src/core/vector/qgsvectorlayertemporalproperties.cpp:450]: Looks ok to me :+1:

**Comment by elpaso** [comment]
@pblottiere , @rldhont can you please have a look to this PR? There a are still a couple of TODOs but I'd like your blessing on the general idea.

**Comment by rldhont** [comment]
Thanks @elpaso I was thinking about doing it but I I didn't take the time to do it.

**Comment by m-kuhn** [comment]
Nice!
Is there a situation where we do not want this? I.e. could the option be removed (or on by default)?

**Comment by elpaso** [comment]
> Nice! Is there a situation where we do not want this? I.e. could the option be removed (or on by default)?

Yes, temporal support can be activate on an individual layer basis and the explicitly set TIME dimension (for vectors) will have the priority anyway so I think we could remove the option.

**Comment by pblottiere** [comment]
Nice, good move @elpaso :tada:!

**Comment by elpaso** [comment]
@m-kuhn option removed, now exposed by default.

**Comment by elpaso** [comment]
@rldhont @pblottiere can I hope for an approval here?

File: src/core/vector/qgsvectorlayertemporalproperties.cpp [hunk 2]
```diff
@@ -442,11 +442,22 @@ QString dateTimeExpressionLiteral( const QDateTime &datetime )
          .arg( datetime.time().second() + datetime.time().msec() / 1000.0 );
 }
 
-QString QgsVectorLayerTemporalProperties::createFilterString( QgsVectorLayerTemporalContext, const QgsDateTimeRange &filterRange ) const
+QString QgsVectorLayerTemporalProperties::createFilterString( QgsVectorLayerTemporalContext context, const QgsDateTimeRange &filterRange ) const
```
Comment (pblottiere) [inline src/core/vector/qgsvectorlayertemporalproperties.cpp:445]: @elpaso do you know why the `context` is not a const ref in this case?

**Comment by pblottiere** [comment]
It looks good to me, thanks @elpaso :+1:

**Review by pblottiere — APPROVED** [review]

File: src/core/vector/qgsvectorlayertemporalproperties.cpp (hunk 2 above)
Comment (elpaso) [reply src/core/vector/qgsvectorlayertemporalproperties.cpp:445]: mmm, no but I'll check if we can make it const ref (or would it be an API change?)

File: src/server/services/wms/qgswmsgetcapabilities.cpp [hunk 3]
```diff
@@ -1282,6 +1292,87 @@ namespace QgsWms
...
+          if ( ! timeDimensionAdded
+               && l->temporalProperties()
+               && l->temporalProperties()->isActive() )
+          {
+
+            QDomElement dimElem = doc.createElement( QStringLiteral( "Dimension" ) );
+            dimElem.setAttribute( QStringLiteral( "name" ), QStringLiteral( "TIME" ) );
+            dimElem.setAttribute( QStringLiteral( "units" ), QStringLiteral( "ISO8601" ) );
+
+            // TODO: set "default" (reference value)
```
[the remark this answers is not in the document]
Comment (elpaso) [reply src/server/services/wms/qgswmsgetcapabilities.cpp:1305]: Not really, I couldn't understand how to get that value and if it is really possible to get it from the temporal API.

File: src/core/vector/qgsvectorlayertemporalproperties.cpp (hunk 2 above)
Comment (elpaso) [reply src/core/vector/qgsvectorlayertemporalproperties.cpp]: done

Comment (pblottiere) [reply src/core/vector/qgsvectorlayertemporalproperties.cpp]: Nice thanks

**Pull request merged**